### PR TITLE
AOVs refactor

### PR DESCRIPTION
**Pull request by est77**
- Simplified AOVs code.
- Geometric AOVs (normals, uvs, depth) are now unfiltered.
- Normal and UV AOVs are saved as RGB images to make them easier to use in comp apps.
- Removed color and alpha multipliers from physical surface shader.

### CODE REVIEW

File: src/appleseed/renderer/modeling/aov/depthaov.cpp [hunk 1]
```diff
@@ -58,34 +64,53 @@ namespace
...
-        virtual void reset() override
+        virtual void on_tile_begin(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y) override
         {
-            m_depth = numeric_limits<float>::max();
+            UnfilteredAOVAccumulator::on_tile_begin(frame, tile_x, tile_y);
+
+            get_tile().clear(Vector2f(std::numeric_limits<float>::max()));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/depthaov.cpp]: Remove `std::` qualifier.

File: src/appleseed/renderer/modeling/aov/depthaov.cpp [hunk 2]
```diff
@@ -58,34 +64,53 @@ namespace
...
+                return;
 
-      private:
-        float m_depth;
+            float* p = reinterpret_cast<float*>(
+                get_tile().pixel(pi.x - m_tile_origin_x, pi.y - m_tile_origin_y));
+
+            const float min_sample_squared_distance = p[1];
+            const float sample_squared_distance =
+                squared_distace_to_pixel_center(pixel_context.get_sample_position());
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/depthaov.cpp]: Typo: *distace* -> *distance*

Also, for the sake of consistency, may I suggest *square* instead of *squared*?
 I.e. `square_distance_to_pixel_center()`

File: src/appleseed/renderer/modeling/aov/depthaov.cpp [hunk 3]
```diff
@@ -31,19 +31,25 @@
...
+#include "foundation/image/image.h"
+#include "foundation/image/tile.h"
 #include "foundation/utility/api/apistring.h"
 #include "foundation/utility/api/specializedapiarrays.h"
 #include "foundation/utility/containers/dictionary.h"
 
 // Standard headers.
 #include <cstddef>
 #include <limits>
+#include <memory>
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/depthaov.cpp]: Why is this header required?

File: src/appleseed/renderer/modeling/aov/diffuseaov.cpp [hunk 4]
```diff
@@ -58,22 +58,31 @@ namespace
...
-            const float                 multiplier) override
+            ShadingResult&              shading_result) override
         {
-            m_color = shading_components.m_diffuse.to_rgb(g_std_lighting_conditions);
-            m_color += shading_components.m_indirect_diffuse.to_rgb(g_std_lighting_conditions);
-            m_color *= multiplier;
+            shading_result.m_aovs[m_index].rgb() =
+                shading_components.m_diffuse.to_rgb(g_std_lighting_conditions);
+
+            shading_result.m_aovs[m_index].rgb() +=
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/diffuseaov.cpp]: For readability and performance, collapse these two lines?

    shading_result.m_aovs[m_index].rgb() =
        shading_components.m_diffuse.to_rgb(g_std_lighting_conditions) +
        shading_components.m_indirect_diffuse.to_rgb(g_std_lighting_conditions);

File: src/appleseed/renderer/modeling/aov/glossyaov.cpp [hunk 5]
```diff
@@ -58,22 +58,31 @@ namespace
...
+            const PixelContext&         pixel_context,
+            const ShadingPoint&         shading_point,
             const ShadingComponents&    shading_components,
-            const float                 multiplier) override
+            ShadingResult&              shading_result) override
         {
-            m_color = shading_components.m_glossy.to_rgb(g_std_lighting_conditions);
-            m_color += shading_components.m_indirect_glossy.to_rgb(g_std_lighting_conditions);
-            m_color *= multiplier;
+            shading_result.m_aovs[m_index].rgb() =
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/glossyaov.cpp]: Same remark.

File: src/appleseed/renderer/modeling/aov/normalaov.cpp [hunk 6]
```diff
@@ -125,7 +147,7 @@ namespace
 
         virtual const char** get_channel_names() const override
         {
-            static const char* ChannelNames[] = {"X", "Y", "Z"};
+            static const char* ChannelNames[] = {"R", "G", "B"};
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/normalaov.cpp]: What's the rationale behind this change?

File: src/appleseed/renderer/modeling/aov/normalaov.cpp [hunk 7]
```diff
@@ -56,40 +59,59 @@ namespace
...
+            float* p = reinterpret_cast<float*>(
+                get_tile().pixel(pi.x - m_tile_origin_x, pi.y - m_tile_origin_y));
+
+            const float min_sample_squared_distance = p[3];
+            const float sample_squared_distance =
+                squared_distace_to_pixel_center(pixel_context.get_sample_position());
+
+            if (sample_squared_distance < min_sample_squared_distance)
+            {
+                const Vector3d& normal =
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/normalaov.cpp]: I'd suggest to get rid of `zero_normal` and simplify the code here:

    if (shading_point_hit())
    {
        const Vector3d& n = shading_point.get_shading_normal();
        p[0] = static_cast<float>(n[0]) * 0.5f + 0.5f;
        p[1] = static_cast<float>(n[1]) * 0.5f + 0.5f;
        p[2] = static_cast<float>(n[2]) * 0.5f + 0.5f;
        p[3] = sample_squared_distance;
    }
    else
    {
        p[0] = 0.5f;
        p[1] = 0.5f;
        p[2] = 0.5f;
        p[3] = sample_squared_distance;
    }

File: src/appleseed/renderer/modeling/aov/normalaov.cpp [hunk 8]
```diff
@@ -56,40 +59,59 @@ namespace
...
+            const Vector2i& pi = pixel_context.get_pixel_coords();
+
+            // Ignore samples outside the tile.
+            if (outside_tile(pi))
+                return;
+
+            float* p = reinterpret_cast<float*>(
+                get_tile().pixel(pi.x - m_tile_origin_x, pi.y - m_tile_origin_y));
+
+            const float min_sample_squared_distance = p[3];
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/normalaov.cpp]: *squared* -> *square*

File: src/appleseed/renderer/modeling/aov/uvaov.cpp [hunk 9]
```diff
@@ -56,35 +59,59 @@ namespace
...
+            float* p = reinterpret_cast<float*>(
+                get_tile().pixel(pi.x - m_tile_origin_x, pi.y - m_tile_origin_y));
+
+            const float min_sample_squared_distance = p[3];
+            const float sample_squared_distance =
+                squared_distace_to_pixel_center(pixel_context.get_sample_position());
+
+            if (sample_squared_distance < min_sample_squared_distance)
+            {
+                const Vector2f& uv =
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/uvaov.cpp]: Same remark here.

File: src/appleseed/renderer/kernel/shading/shadingresult.h [hunk 10]
```diff
@@ -72,6 +72,8 @@ class ShadingResult
 
     // Apply alpha premultiplication to the main output and the AOVs.
     void apply_alpha_premult();
+
+    void set_to_opaque_pink();
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingresult.h]: `set_main_to_opaque_pink()` since it does not touch AOVs (in particular it doesn't reset `m_aov_count`).

File: src/appleseed/renderer/kernel/rendering/generic/generictilerenderer.cpp [hunk 11]
```diff
@@ -155,6 +155,9 @@ namespace
             // Inform the pixel renderer that we are about to render a tile.
             m_pixel_renderer->on_tile_begin(frame, tile, aov_tiles);
 
+            // Inform the aov accumulators that we are about to render a tile.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/generic/generictilerenderer.cpp]: *aov* -> *AOV*

File: src/appleseed/renderer/kernel/aov/aovaccumulator.h [hunk 12]
```diff
@@ -65,110 +69,82 @@ class AOVAccumulator
...
-    virtual void reset() override;
+    return *m_tile;
+}
 
-    virtual void flush(ShadingResult& result) override;
+inline bool UnfilteredAOVAccumulator::outside_tile(
+    const foundation::Vector2i& pi) const
+{
+   return
+    pi.x < m_tile_origin_x || pi.y < m_tile_origin_y ||
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/aov/aovaccumulator.h]: Fix indentation.

File: src/appleseed/renderer/modeling/project/projectfileupdater.cpp [hunk 13]
```diff
@@ -1510,7 +1510,10 @@ namespace
         virtual void update() override
         {
             if (Scene* scene = m_project.get_scene())
+            {
                 update_material_and_object_inputs(scene->assemblies());
+                update_physical_surface_shader_inputs(scene->assemblies());
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/projectfileupdater.cpp]: Please increase the project file format revision.

File: src/appleseed/renderer/modeling/aov/aov.cpp [hunk 14]
```diff
@@ -87,4 +113,40 @@ bool ColorAOV::has_color_data() const
...
+//
+
+UnfilteredAOV::UnfilteredAOV(const char* name, const ParamArray& params)
+  : AOV(name, params)
+{
+}
+
+UnfilteredAOV::~UnfilteredAOV()
+{
+    delete m_image;
```
[the remark this answers is not in the document]
Comment (est77) [reply src/appleseed/renderer/modeling/aov/aov.cpp]: For unfiltered AOVs, images are created here:
https://github.com/appleseedhq/appleseed/pull/1641/files/e457f97da0b0a6b5e745e5020707cdeec5bc6888#diff-231966861e8dda0582dc2510e8165c38R143

File: src/appleseed/renderer/modeling/aov/normalaov.cpp (hunk 6 above)
Comment (est77) [reply src/appleseed/renderer/modeling/aov/normalaov.cpp]: It makes it much easier to use the resulting images in compositing applications.
